src/ruby/ext/grpc/rb_xds_channel_credentials.c

```
@@ -0,0 +1,212 @@
+     grpc_rb_xds_channel_credentials_free,
+     GRPC_RB_MEMSIZE_UNAVAILABLE,
+     NULL},
+    NULL,
+    NULL,
+#ifdef RUBY_TYPED_FREE_IMMEDIATELY
+    RUBY_TYPED_FREE_IMMEDIATELY
+#endif
+};
+
+/* Allocates ChannelCredential instances.
+   Provides safe initial defaults for the instance fields. */
+static VALUE grpc_rb_xds_channel_credentials_alloc(VALUE cls) {
+  grpc_ruby_init();
+  grpc_rb_xds_channel_credentials* wrapper =
+      ALLOC(grpc_rb_xds_channel_credentials);
+  wrapper->wrapped = NULL;
+  wrapper->mark = Qnil;
+  return TypedData_Wrap_Struct(cls, &grpc_rb_xds_channel_credentials_data_type,
+                               wrapper);
+}
+
+/* Creates a wrapping object for a given channel credentials. This should only
+ * be called with grpc_channel_credentials objects that are not already
+ * associated with any Ruby object. */
+VALUE grpc_rb_xds_wrap_channel_credentials(grpc_channel_credentials* c,
+                                       VALUE mark) {
+  VALUE rb_wrapper;
+  grpc_rb_xds_channel_credentials* wrapper;
+  if (c == NULL) {
+    return Qnil;
+  }
+  rb_wrapper =
+      grpc_rb_xds_channel_credentials_alloc(grpc_rb_cXdsChannelCredentials);
+  TypedData_Get_Struct(rb_wrapper, grpc_rb_xds_channel_credentials,
+                       &grpc_rb_xds_channel_credentials_data_type, wrapper);
+  wrapper->wrapped = c;
+  wrapper->mark = mark;
+  return rb_wrapper;
+}
+
+/* The attribute used on the mark object to hold the fallback creds. */
+static ID id_fallback_creds;
+
+/*
+  call-seq:
+    fallback_creds: (ChannelCredentials) fallback credentials to create
+    XDS credentials
+    Initializes Credential instances. */
+static VALUE grpc_rb_xds_channel_credentials_init(VALUE self,
+                                                  VALUE fallback_creds) {
+  grpc_rb_xds_channel_credentials* wrapper = NULL;
+  grpc_channel_credentials* grpc_fallback_creds =
+      grpc_rb_get_wrapped_channel_credentials(fallback_creds);
```

nit: can we check if `fallback_creds` is `nil` before calling this?

That way, we can throw an error which gives more specific details than the catch-call `"could not create a credentials, not sure why"` error below

---

Same as previous, there are tests passing nil and non-ChannelCredentials object